Adding new zero-shot examples

SangbumChoi opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes https://github.com/huggingface/transformers/issues/32459

Following two PR should be merged before merging this example!
https://github.com/huggingface/transformers/pull/31828
https://github.com/huggingface/transformers/pull/31964

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

SangbumChoi commented:
@qubvel Hi pavel! Is there any standard batch size that hugginface suggest? The reason why I am asking is unlike standard detr, GroundingDino cannot handle batch size 8 for normal single GPU.

SangbumChoi commented:
@qubvel Hi Pavel this PR is ready for review. However, there is one problem with using evaluation process in `Trainer`.

In order to use the `Trainer` from the transformers we have to get `input_ids` for the post_processing pipeline which is from the tokenizer output. However, transformers `TrainingArgument` `include_input_for_metric` is set to get input variable of first index of class variable which is `pixel_values`. This is hard-coded so I cannot get the input_ids for the input. Also cc @muellerzr.

https://github.com/huggingface/transformers/blob/aefd3e2ae1e2ae5ae555028c0409118bae9e12ee/src/transformers/trainer.py#L3865

There is three ways for this
1. having additional argument to set the name
2. change the order of `forward()`
3. change to main_input_name = getattr(self.model, "input_ids", "input_ids")
WDYT?

qubvel commented:
Hi @SangbumChoi, thanks for working on this!

> Is there any standard batch size that hugginface suggest? (In my experience maximum batch size was 2 with 24GB) The reason why I am asking is unlike standard detr, GroundingDino cannot handle batch size 8 for normal single GPU.
I think it is fine to have batch size 2 if the model can be trained with such a batch size. In README and a comment we can mention how much memory one would require to run the script and also advise using gradient accumulation, lower precision, and multi-GPU setup (if it is supported).

qubvel commented:
> However, there is one problem with using evaluation process in Trainer.
In order to use the Trainer from the transformers we have to get input_ids for the post_processing pipeline which is from the tokenizer output. However, transformers TrainingArgument include_input_for_metric is set to get input variable of first index of class variable which is pixel_values. This is hard-coded so I cannot get the input_ids for the input. 

We can probably move these lines
```python
main_input_name = getattr(self.model, "main_input_name", "input_ids")
inputs_decode = self._prepare_input(inputs[main_input_name]) if args.include_inputs_for_metrics else None
```
to a separate method of `Trainer`, for example, `_get_inputs(...)`. Then we will be able to override it in a new `ZeroShotTrainer` class in a script to return required inputs.

muellerzr commented:
agreed re; `get_inputs`

SangbumChoi commented:
@qubvel Hi requesting for first review. I will state changes from the original object_detection.py

1. Instead of wandb logging graph, I just added table of GPU consumption (since there is no other architecture to compare)
2. Also as you might know calculating mAP from zero-shot is not accurate. Currently I have made workaround which is the logic that if the output text is directly in `label2id` then match otherwise treat as background object as 0. (We can discuss this if it seems inappropriate). See `def convert_zero_shot_to_coco_format`
3. Added ZeroShotTrainer to override Trainer with `def _get_input_by_name`

Also kudos to @EduardoPach for `backbone_freeze` and `text_backbone_freeze`. Which make the output more stable :)

Before finetune
<img width="1113" alt="Screenshot 2024-08-09 at 2 15 04 PM" src="https://github.com/user-attachments/assets/fc8c8cc6-e426-4f1d-a49b-fa5ec03ec7d2">
After finetune
<img width="1118" alt="Screenshot 2024-08-09 at 2 15 15 PM" src="https://github.com/user-attachments/assets/57bd8ec4-0645-45ca-8bbf-fb9d1ea19119">

qubvel commented on src/transformers/trainer.py:
```diff
@@ -4452,7 +4456,7 @@ def prediction_loop(
             loss, logits, labels = self.prediction_step(model, inputs, prediction_loss_only, ignore_keys=ignore_keys)
-            main_input_name = getattr(self.model, "main_input_name", "input_ids")
+            main_input_name = self._get_input_by_name()
             inputs_decode = self._prepare_input(inputs[main_input_name]) if args.include_inputs_for_metrics else None
```
Lets have both lines moved to a separate method to be able to select even multiple input keys

```python
def _select_inputs_for_validation(inputs):
    main_input_name = getattr(self.model, "main_input_name", "input_ids")
    return inputs[main_input_name]

...
   selected_inputs = _select_inputs_for_validation(inputs)
   inputs_decode = self._prepare_input(selected_inputs) if args.include_inputs_for_metrics else None
```

In ZeroShotTrainer

```python
class ZeroShotTrainer(Trainer):
    def _select_inputs_for_validation(inputs):
        return inputs["input_ids"]
```

In another trainer
```python
class CustomTrainer(Trainer):
    def _select_inputs_for_validation(inputs):
        return {"input_ids": inputs["input_ids"], "pixel_values": inputs["pixel_values"]}
```

qubvel reviewed (commented):
Hi @SangbumChoi, thanks for iterating! Great work!

I left a comment regarding Trainer, also don't know if it deserves a separate PR. Overall looks very good to me, but I didn't do an in-depth view yet (let's wait for loss PR to be merged). For these examples the tests have to be added and information in README (see files in this [directory](https://github.com/huggingface/transformers/tree/main/examples/pytorch)). Also, feel free to add resources to docs of Grounding DINO. Btw, did you try to fine-tune OwlV2?

SangbumChoi commented:
@qubvel Sounds good. We can wait until the loss PR get merged.
> Btw, did you try to fine-tune OwlV2?
I didn't aware that we can fine-tune OwlV2, let me see while waiting loss PR!

EduardoPach commented on examples/pytorch/zero-shot/run_zero_shot_object_detection.py:
```diff
@@ -0,0 +1,637 @@
+        post_processed_output = processor.post_process_grounded_object_detection(
+            output, input_ids, box_threshold=box_threshold, text_threshold=text_threshold, target_sizes=target_sizes
+        )
+        post_processed_output = convert_zero_shot_to_coco_format(post_processed_output, label2id)
```
@SangbumChoi Not sure if this will still work if `random_text_promt=True` because the `class_labels` from targets will not necessarily follow `label2id`

EduardoPach commented on examples/pytorch/zero-shot/run_zero_shot_object_detection.py:
```diff
@@ -0,0 +1,637 @@
+        labels = prediction["labels"]
+        for label in labels:
+            if label in label2id:
+                torch_label.append(label)
```
I believe you meant `torch_label.append(label2id[label])` right?

SangbumChoi commented:
Hi @qubvel It is now ready to review this new example. For the CI error it says `unauthorized` I'm not sure how to handle this so it would be appreciated to let me know how to fix this.

SangbumChoi commented:
<img width="1447" alt="Screenshot 2024-09-27 at 4 36 25 PM" src="https://github.com/user-attachments/assets/89ce9d7c-5076-409d-ac38-593d74db259c">

junxnone commented:
> <img alt="Screenshot 2024-09-27 at 4 36 25 PM" width="1447" src="https://private-user-images.githubusercontent.com/34004152/371439481-89ce9d7c-5076-409d-ac38-593d74db259c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDkzODQsIm5iZiI6MTczMjg0OTA4NCwicGF0aCI6Ii8zNDAwNDE1Mi8zNzE0Mzk0ODEtODljZTlkN2MtNTA3Ni00MDlkLWFjMzgtNTkzZDc0ZGIyNTljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI5VDAyNTgwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0YmM3MDQ0NjQ5NmFjYTk0YjcxMmM1MTZjNDE0MzcxNzM1NDNlNzc1MTIxYzQ0NmNlYjJhOWMyZDNlODQ5N2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XYK2e6G1_QXaLhqCCIaPNnruKLI8I2vFDsizAdkoiME">

Thanks for share the finetune code.

I test with the code https://github.com/SangbumChoi/transformers on branch grounding_examples.
and only get the lower MAP, it is same as the tensorboard logs in your model [here](https://huggingface.co/danelcsb/grounding-dino-tiny-finetuned-cppe-5-10k-steps-no-trainer/tensorboard)

![image](https://github.com/user-attachments/assets/8a60696b-15e1-4fbb-99c0-34ae4e49ec16)

Looks like I missed something...

SangbumChoi commented:
@junxnone Thanks for the interest. This is because due to difference between zero-shot to coco converted AP. I have fixed in my code but updated in here :(

junxnone commented on examples/pytorch/zero-shot/run_zero_shot_object_detection_no_trainer.py:
```diff
@@ -0,0 +1,895 @@
+    return boxes
+
+
+def convert_zero_shot_to_coco_format(predictions, label2id):
+    """
+    Convert zershot format output to typical object detection format in order to calculate mAP.
+
+    Args:
+        predictions (Dict): Output of zero-shot object detection
+            e.g. {'scores': tensor([0.4786, 0.4379, 0.4760], device='cuda:0'), 'labels': ['a cat', 'a cat', 'a remote control'], 'boxes': tensor([[344.6973,  23.1085, 637.1817, 374.2748],[ 12.2690,  51.9104, 316.8564, 472.4341],[ 38.5870,  70.0092, 176.7755, 118.1748]], device='cuda:0')}
+        label2id (Dict): Dictionary of label to id mapping
+
+    Returns:
+        Dict: Output of zero-shot object detection
+            e.g. {'scores': tensor([0.4786, 0.4379, 0.4760], device='cuda:0'), 'labels': [1, 1, 2], 'boxes': tensor([[344.6973,  23.1085, 637.1817, 374.2748],[ 12.2690,  51.9104, 316.8564, 472.4341],[ 38.5870,  70.0092, 176.7755, 118.1748]], device='cuda:0')}
+
+    """
+    # convert center to corners format
+    torch_label = []
+    for prediction in predictions:
+        scores = prediction["scores"]
+        device = scores.device
+        labels = prediction["labels"]
+        for label in labels:
+            if label in label2id:
+                torch_label.append(label2id[label])
+            else:
+                # Give background class
+                torch_label.append(0)
+        prediction["labels"] = torch.Tensor(torch_label).to(dtype=torch.int32).to(device)
+
+    return predictions
```
I also checkout the pr, and got the same result.

SangbumChoi replied:
I was meaning I have my own code which is not commited here

junxnone commented on examples/pytorch/zero-shot/run_zero_shot_object_detection.py:
```diff
@@ -0,0 +1,637 @@
+        label2id=label2id,
+        random_text_prompt=False,
+    )
+    validation_transform_batch = partial(
+        augment_and_transform_batch,
+        transform=validation_transform,
+        processor=processor,
+        id2label=id2label,
+        label2id=label2id,
+        random_text_prompt=True,
+    )
```
I have some confusion here. Why with random_text_prompt=True do I get bad results?

SangbumChoi replied:
This will be deprecated. You understand correctly

JamesEmi replied:
@SangbumChoi thanks a ton for your work so far! I am looking to use this code to finetune GroundingDINO for a critical medical imaging project - and from my understanding of the thread here I am noting some issues I should be careful about. Do you have any points on 'things to remember' when using this code? Or if you have a latest clean version would be super useful to us if you can commit those updates. One other option I am considering is finetuning GroundingDINO using [this repo](https://github.com/longzw1997/Open-GroundingDino); but would still prefer finetuning the model off `transformers`. Thanks!!

JamesEmi commented:
@qubvel @EduardoPach thanks for your work on this so far (and in general for the wider HF ecosystem :) ). Ended up here after digging around for a train script entrypoint to use for [finetuning grounding dino tiny](https://huggingface.co/IDEA-Research/grounding-dino-tiny). Any updates on why this PR is not merged to the main branch? @EduardoPach could you comment on what training script you used to finetune grounding dino tiny to get [this model](https://huggingface.co/EduardoPacheco/grounding-dino-tiny-aquarium-fine-tune)? Thanks in advance!!!

sbchoi-droid committed:
Add Grounding DINO zero-shot object detection fine-tuning examples

Port the zero-shot object detection (Grounding DINO) fine-tuning example
scripts onto current main and update them for the current Transformers API:

- run_zero_shot_object_detection.py (Trainer) and
  run_zero_shot_object_detection_no_trainer.py (Accelerate), plus README
  and requirements.
- Trainer script: drop the obsolete ZeroShotTrainer._select_inputs_for_validation
  hook, use processing_class instead of the deprecated tokenizer argument, and
  rebuild input_ids from a deterministic eval prompt inside compute_metrics
  (the Trainer no longer surfaces input_ids via include_for_metrics).
- Rename post_process_grounded_object_detection box_threshold -> threshold.
- README: eval_strategy (was evaluation_strategy), remove include_inputs_for_metrics,
  fix inference snippet and script link.
- Register the example in examples/pytorch/README.md.
- Bump check_min_version to current.

sbchoi-droid committed:
docs: note MM Grounding DINO support in zero-shot object detection example

The zero-shot object detection scripts work out of the box with the Grounding
DINO family (Grounding DINO and MM Grounding DINO), which share the same
processor and detection loss. Document this with an MM Grounding DINO run
command and clarify that OWL-ViT/OWLv2/OmDet-Turbo are not supported yet
(they don't implement an object detection training loss).

sbchoi-droid committed:
Add object detection training loss to OWL-ViT and OWLv2

OWL-ViT and OWLv2 could only be used for inference because their
`ForObjectDetection` forward never accepted `labels` nor computed a loss.
Their classification logits are shaped (batch, num_patches, num_queries) =
one logit per predicted box per text query, which is structurally a
DeformableDETR-style sigmoid detector, so the loss reuses the existing
Hungarian matcher and image loss (sigmoid focal classification + L1 + GIoU).

- Add `OwlViTForObjectDetectionLoss` (src/transformers/loss/loss_owlvit.py),
  deriving the number of classes from the logits (open-vocabulary, not fixed
  by config), and register it for both models in LOSS_MAPPING.
- Add loss hyperparameters (class/bbox/giou costs, loss coefficients,
  focal_alpha) to OwlViTConfig (propagated to Owlv2Config via Copied from).
- Wire `labels` + loss computation into both forwards; populate the existing
  loss/loss_dict output fields.
- Allow the new loss-only config attributes in check_config_attributes.
- Add a test_object_detection_loss test (loss is a scalar, has the expected
  components, gradients flow) for both models.

This makes OWL-ViT/OWLv2 fine-tunable, e.g. via the zero-shot object
detection example.

sbchoi-droid committed:
Add object detection training loss to OmDet-Turbo

OmDet-Turbo previously raised NotImplementedError when `labels` were passed.
It is a DETR-style (RT-DETR-like) detector predicting, per decoder query, a box
and one logit per candidate class, so the loss reuses the DeformableDETR-style
matcher and image loss (sigmoid focal classification + L1 + GIoU). The number of
classes is open-vocabulary, so it is read from the logits. When
`config.auxiliary_loss` is enabled, a loss is computed at every intermediate
decoder layer (the decoder stacks per-layer class/box predictions in training).

- Add `OmDetTurboForObjectDetectionLoss` (src/transformers/loss/loss_omdet_turbo.py)
  and register it in LOSS_MAPPING.
- Add loss hyperparameters (auxiliary_loss, class/bbox/giou costs, loss
  coefficients, focal_alpha) to OmDetTurboConfig.
- Replace the NotImplementedError with the actual loss computation; populate the
  output `loss` and a new `loss_dict` field.
- Allow the new loss-only config attributes in check_config_attributes.
- Add a test_object_detection_loss test (scalar loss, expected components incl.
  auxiliary losses, gradients flow).

With this, all three previously inference-only zero-shot detectors (OWL-ViT,
OWLv2, OmDet-Turbo) can be fine-tuned.

sbchoi-droid committed:
Fix OmDet-Turbo language cache breaking training

The language embedding LRU cache stores graph-bearing tensors and reuses them
across forward passes. During training this both reuses already-freed autograd
graphs (RuntimeError: backward through the graph a second time) and stops
gradients from reaching the language backbone. Bypass the cache when
self.training is True so embeddings are recomputed fresh each step; inference
behaviour is unchanged.

Verified with an overfit sanity check: the OmDet-Turbo train loss now decreases
steadily (87.4 -> 15.5 over 80 steps), as do OWL-ViT, OWLv2 and Grounding DINO.

sbchoi-droid committed:
Add zero-shot detector fine-tuning benchmark script

Self-contained script to fine-tune and compare the zero-shot object detectors
that now support a training loss (Grounding DINO, OWL-ViT, OWLv2, OmDet-Turbo)
on a small/sub-sampled detection dataset. It logs train loss, validation loss
and validation mAP per epoch, renders vanilla-vs-fine-tuned sample inference
images, writes metrics.json, and can push the artifacts to the Hub. Used to
verify the loss decreases and to compare models; also runnable via HF Jobs.

sbchoi-droid committed:
Fix device mismatch in benchmark sample rendering on GPU

Grounding DINO predicted label ids were built on CPU while boxes/scores were on
the model device, breaking masked indexing during sample rendering on CUDA.
Build label ids on the boxes' device and move predictions to CPU before drawing.

sbchoi-droid committed:
Improve zero-shot fine-tuning benchmark: correct AP eval + frozen backbones

Deep-analysis fixes after the first PCB run showed flat val loss and ~0 AP:
- Evaluate via each processor's post_process_* with target_sizes, so predicted
  boxes are mapped back to the original image and labels are proper class indices
  (the previous manual conversion mis-handled resize/padding, and Grounding DINO's
  multi-word class names never matched -> mAP was 0).
- Freeze image+text backbones by default (train only the detection heads / encoder-
  decoder) to curb over-fitting on small datasets and speed up training.
- Build Grounding DINO training labels from the processor's COCO annotations (correct
  box frame); add horizontal-flip augmentation; support dropping RF100 meta-class 0
  and cleaning class names.
- Default to an out-of-domain industrial dataset (Francesco/excavators-czvg9).

sbchoi-droid committed:
Use manual labels for Grounding DINO too (processor lacks annotations kwarg)

The Grounding DINO processor doesn't accept an `annotations` argument, so build
its training labels manually like the other models; normalized cxcywh boxes are
valid because Grounding DINO uses aspect-preserving resize with no padding at
batch size 1.

sbchoi-droid committed:
Fix Grounding DINO eval: use non-zero text_threshold for label decoding

With text_threshold=0.0 the phrase extraction returns every text token, so each box
decodes to the whole prompt and all predictions collapse onto a single class (mAP ~0).
Use threshold=0.05 / text_threshold=0.25 for the metric computation.

sbchoi-droid committed:
Keep Grounding DINO eval predictions length-consistent

post_process can return a phrase list whose length differs from the boxes (e.g. 0 boxes
but a stray label), which torchmetrics rejects. Align boxes/scores/labels to the number
of boxes.

sbchoi-droid committed:
Log vanilla (zero-shot) AP baseline + add early stopping to benchmark

Evaluate the pretrained model before any training (epoch -1) so the true zero-shot AP
is recorded, and stop training when val mAP stalls (patience). Record vanilla_map /
best_map / best_epoch in the summary.

SangbumChoi committed:
Merge pull request #8 from SangbumChoi/claude/branch-name-grounding-underbar-72aLc

Make zero-shot object detectors fine-tunable (OWL-ViT, OWLv2, OmDet-Turbo) + zero-shot OD examples & benchmark

SangbumChoi committed:
Merge branch 'main' into grounding_examples

SangbumChoi committed:
make fix

SangbumChoi commented:
@yonigozlan @molbap anybody interested in this PR?

SangbumChoi commented:
https://github.com/SangbumChoi/transformers/pull/8#issuecomment-4644749929

SangbumChoi committed:
Merge branch 'huggingface:main' into grounding_examples

SangbumChoi committed:
Merge branch 'main' into grounding_examples
